File: source/compiler.cpp
#include "compiler.hpp"
#include "parser.hpp"
#include "linker.hpp"
#include "config.hpp"

// Compile
bool compile(void)
{
    Parser parser;
    bool   had_error = false;

    // Parse source code
    if (!parser.start())
        return false;

    for (Input &input : g_inputs)
    {
        // Semantic analysis
        if (!input.analyze())
        {
            had_error = true;
            continue;
        }

        // Generate C++
        input.generate();
    }

    // An error happened?
    if (had_error)
        return false;

    // Link code
    if (!link())
        return false;

    return true;
}

File: source/Compiler.cs
namespace Suteki
{
    class Compiler
    {
        // Start the compiler
        public static void Start()
        {
            // Parse source code
            Parser parser = new Parser();
            parser.Start();

            // Only emit if no error
            if (Config.HadError)
                return;

            // Emit C++ code
            foreach (Input input in Config.Inputs)
            {
                foreach (Node node in input.Nodes)
                {
                    node.Emit(input);
                }
            }
        }
    }
}